routing/index_graph.cpp

```diff
@@ -1,10 +1,115 @@
+
+  // Looking for at least one restriction of type No from |featureIdFrom| to |featureIdTo|.
+  if (binary_search(restrictions.cbegin(), restrictions.cend(),
+                    Restriction(Restriction::Type::No, {featureIdFrom, featureIdTo})))
+  {
+    if (featureIdFrom != featureIdTo)
+      return true;
+
+    // @TODO(bykoianko) According to current code if a feature id is marked as a feature with
+    // restrictricted uturn it's restricted to make a uturn on the both ends of the feature.
+    // Generally speaking it's wrong. In osm there's information about the end of the feature
+    // where the uturn is restricted. It's necessary to pass the data to mwm and to use it here.
+    // Set test LineGraph_RestrictionF1F1No for further details.
+    return u.GetSegmentIdx() == v.GetSegmentIdx() && u.IsForward() != v.IsForward();
+  }
+
+  // Taking into account that number of features starting and ending with the same feature id
+  // is relevantly small it's possible to ignore such cases.
+  if (featureIdFrom == featureIdTo)
+    return false;
+
+  // Looking for at least one restrictions of type Only from |featureIdFrom| to |featureIdTo|.
+  if (binary_search(restrictions.cbegin(), restrictions.cend(),
+                    Restriction(Restriction::Type::Only, {featureIdFrom, featureIdTo})))
+  {
+    return false;
+  }
+
+  // Looking for at least one restriction of type Only starting from |featureIdFrom|
+  // and with finish different from |featureIdTo|.
+  // Note. There is the following idea behind the binary_search below
+  // * if there is an item of type Only in |restrictions| starting from |featureIdFrom|
+  //   and finishing with |featureIdTo| the method returns false earlier.
+  // * binary_search below looks for a restriction of type Only starting from |featureIdFrom|.
+  //   If such restriction is found it means there is a restiction in |restrictions| of type Only
+  //   starting from |featureIdFrom| and finishing at feature id different from |featureIdTo|.
+  //   So it's impossible to move from |featureIdFrom| to |featureIdTo|.
+  auto const it = lower_bound(restrictions.cbegin(), restrictions.cend(),
+                              Restriction(Restriction::Type::Only, {featureIdFrom, featureIdTo}),
+                              [](Restriction const & r1, Restriction const & r2) {
+                                CHECK(!r1.m_featureIds.empty(), ());
+                                CHECK(!r2.m_featureIds.empty(), ());
+                                if (r1.m_type != r2.m_type)
+                                  return r1.m_type < r2.m_type;
+                                return r1.m_featureIds[0] < r2.m_featureIds[0];
+                              });
+
+  if (it == restrictions.cend() || it->m_type != Restriction::Type::Only ||
+      it->m_featureIds[0] != featureIdFrom)
+    return false;
+
+  // Note. At this point it's clear that there is an item in |restrictions| with type Only
```

Comment:
Why do we need to check edges? Why not to check that restriction vector contains restriction of type Only, with the first component equal to featureIdFrom and the second component NOT equal to featureIdTo? And, if so, we need to return true from IsRestricted.